src/App/GroupExtension.cpp

```diff
@@ -428,29 +421,97 @@ void GroupExtension::extensionOnChanged(const Property* p) {
+            // its view provider) to inform us that we should not toggle
+            // children visibility here.
+            _checkParentGroup = _checkParentGroup 
+                && GeoFeatureGroupExtension::getGroupOfObject(owner);
+
+            auto hiddenChildren = Base::freecad_dynamic_cast<PropertyMap>(
+                    owner->getPropertyByName("HiddenChildren"));
+            if(hiddenChildren && hiddenChildren->getContainer()!=owner)
+                hiddenChildren = 0;
+
             Base::FlagToggler<> guard(_togglingVisibility);
+
+            std::map<std::string,std::string> hc;
+
             for(auto obj : Group.getValues()) {
-                if(obj && obj->getNameInDocument() && obj->Visibility.getValue()!=vis) {
-                    touched = true;
-                    obj->Visibility.setValue(vis);
-                }
+                if(!obj || !obj->getNameInDocument())
+                    continue;
+                if(obj->Visibility.getValue()!=vis) {
+                    if(vis && hiddenChildren && hiddenChildren->getValue(obj->getNameInDocument()))
+                        continue;
+                    if(!_checkParentGroup) {
+                        touched = true;
+                        obj->Visibility.setValue(vis);
+                    }
+                } else if (!vis && hiddenChildren)
+                    hc.emplace(obj->getNameInDocument(),"");
+            }
+
+            if(!vis && hiddenChildren) {
+                // If all children are invisible, do not keep the record, so
+                // that next time this group is shown, all its children will be
+                // shown.
+                if(hc.size() == Group.getValues().size())
+                    hc.clear();
+                hiddenChildren->setValues(std::move(hc));
             }
+
             if(touched) {
-                Base::ObjectStatusLocker<Property::Status, Property> guard(Property::Output, &_GroupTouched);
-                // Temporary set the Property::Output on _GroupTouched, so that
-                // it does not touch the owner object, but still signal
-                // Part::Feature shape cache update.
-                _GroupTouched.touch();
+                if(_GroupTouched.testStatus(Property::Output))
+                    _GroupTouched.touch();
+                else {
+                    Base::ObjectStatusLocker<Property::Status, Property> guard(Property::Output, &_GroupTouched);
+                    // Temporary set the Property::Output on _GroupTouched, so that
+                    // it does not touch the owner object, but still signal
+                    // Part::Feature shape cache update.
+                    _GroupTouched.touch();
+                }
             }
         }
     }
 
     App::Extension::extensionOnChanged(p);
 }
 
-void GroupExtension::slotChildChanged(const DocumentObject &obj, const Property &prop) {
-    if(&prop == &obj.Visibility && !_togglingVisibility)

```

Comment:
Reading this code, I wonder: Shouldn't this `&prop == &obj.Visibility` be kept? I'm not entirely sure, but it looks like the below handler only has code for when a child's visibility changes? Or is this changed handler somehow limited to only Visibility changes now?

Reply:
Yes, it is limited to Visibility. A new signalChanged has been added to every property. This `slotChildChanged` is now hooked to child objects' `Visibility.signalChanged`, so no need to check for property anymore.

Reply:
Ah, I see it was in an earlier commit, I was looking just in the same commit. Ok, looks good, thanks for clarifying!